segmentation.py

```diff
@@ -0,0 +1,575 @@
+        # hypotheses.
+        if self.__alg__ == 'maxprod':
+            loglik = logmax
+            self.__index.append(ind)
+
+        # In the sum-product algorithm, keep track of the probabilities.
+        else:
+            iteration = list()
+            for hypothesis in self.__hypotheses:
+                iteration.append((hypothesis['count'],
+                                  hypothesis['log_probability']))
+
+            self.__probabilities.append(iteration)
+
+    def trim(self, minprob=1.0e-6, maxhypot=20):
+        """Prune hypotheses to limit computational complexity.
+
+        The computational complexity of the algorithm can be managed by
+        limiting the number of hypotheses maintained. This method limits the
+        number of hypotheses maintained by the model by:
+
+            1) Removing any hypotheses with a support (probability) less than
+               ``minprob``.
+
+            2) Preserving the first ``maxhypot`` likely hypotheses and
+               discarding the rest.
+
+        """
+
+        # Skip pruning if less hypotheses exist than the maximum allowed.
+        if len(self.__hypotheses) <= maxhypot:
+            return
+
+        # Sort the hypotheses in decreasing log probability order.
+        self.__hypotheses.sort(key=lambda dct: -dct['log_probability'])
+
+        # Store the indices of likely hypotheses.
+        minprob = np.log(minprob)
+        index = [i for i, hypot in enumerate(self.__hypotheses)
+                 if hypot['log_probability'] > minprob]
+
+        # Trim the hypotheses.
+        index = index[:maxhypot] if len(index) >= maxhypot else index
+        self.__hypotheses = [self.__hypotheses[i] for i in index]
+
+        # NOTE: This final ordering can preserve the original order of the
+        #       hypotheses. Interestingly, the algorithm specified in update
+        #       does not require that the hypotheses be ordered! This sort can
+        #       safely be ignored.
+        # self.__hypotheses.sort(key=lambda dct: dct['index'])
+
+        # Normalize the hypotheses so that their probabilities sum to one.
+        logsum = -np.inf
+        for hypot in self.__hypotheses:
+            logsum = self.__accum(logsum, hypot['log_probability'])
+        for hypot in self.__hypotheses:
+            hypot['log_probability'] -= logsum
+
+    def segment(self):
+        """Return posterior probabilities OR sequence segmentation.
+
+        If the MAX-PRODUCT algorithm is selected, this method returns the most
+        likely sequence segmentation as a list of integers. Each integer in the
+        list marks where a segment begins.
+
+        If the SUM-PRODUCT algorithm is selected, this method returns the
+        posterior probabilities of the segmentation hypotheses as a numpy
+        array. Rows in the array represent hypotheses and columns in the array
+        represent data points in the time-series.
+
+        Returns:
```

Comment:
This method returns the inference results In the case of the max-product algorithm, the method returns the most likely segmentation ...